### PR TITLE
Feature/http executor

### PR DESCRIPTION
**Pull request by bostko**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,6 +5,9 @@ notifications:
     on_success: change
     on_failure: always
 
+before_install:
+  - sed -i.bak -e 's|https://nexus.codehaus.org/snapshots/|https://oss.sonatype.org/content/repositories/codehaus-snapshots/|g' ~/.m2/settings.xml
```
Comment (bostko) [inline .travis.yml:9]: So you need snapshots for the build.
What is the snapshot dependency you are trying to get? Is it Apache Brooklyn?

Comment (ygy) [reply .travis.yml:9]: nexus.codehaus.org is no longer available (codehaus has been shutdown).
But this address is still used by Travis, so it needs to be replaced

File: ambari/src/main/java/io/brooklyn/ambari/rest/AmbariRestClient.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+import retrofit.client.Client;
+import retrofit.client.Header;
+import retrofit.client.Request;
+import retrofit.client.Response;
+import retrofit.client.UrlConnectionClient;
+import retrofit.mime.TypedByteArray;
+import retrofit.mime.TypedInput;
+import retrofit.mime.TypedOutput;
+
+public class AmbariRestClient implements Client {
```
Comment (duncangrant) [inline ambari/src/main/java/io/brooklyn/ambari/rest/AmbariRestClient.java:54]: I think we need to document what this is for

Comment (ygy) [reply ambari/src/main/java/io/brooklyn/ambari/rest/AmbariRestClient.java:54]: Docs have been added.

File: ambari/src/main/java/io/brooklyn/ambari/rest/AmbariRestClient.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+import retrofit.client.Header;
+import retrofit.client.Request;
+import retrofit.client.Response;
+import retrofit.client.UrlConnectionClient;
+import retrofit.mime.TypedByteArray;
+import retrofit.mime.TypedInput;
+import retrofit.mime.TypedOutput;
+
+/**
+ * Represent a REST client which provides the ability to use different HttpExecetor implementations as defined by
```
Comment (duncangrant) [inline ambari/src/main/java/io/brooklyn/ambari/rest/AmbariRestClient.java]: httpexecutor

File: ambari/src/main/java/io/brooklyn/ambari/rest/AmbariRestClient.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                byte[] content = null;
+                TypedOutput body = request.getBody();
+                if (body != null) {
+                    mimeType = request.getBody().mimeType();
+                    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                    try {
+                        body.writeTo(baos);
+                        content = baos.toByteArray();
+                    } catch (IOException e) {
+                        throw e;
```
Comment (duncangrant) [inline ambari/src/main/java/io/brooklyn/ambari/rest/AmbariRestClient.java]: Any reason not to remove catch clause?

File: ambari/src/main/java/io/brooklyn/ambari/rest/AmbariRestClient.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+/**
+ * Represent a REST client which provides the ability to use different HttpExecutor implementations as defined by
+ * {@link org.apache.brooklyn.util.http.executor.HttpExecutor} or produced by {@link org.apache.brooklyn.util.executor.HttpExecutorFactory}.
+ *
+ * It uses the Retrofit library default client, if factory is not set.
+ */
+public class AmbariRestClient implements Client {
+
+    private static Client client;
```
Comment (bostko) [inline ambari/src/main/java/io/brooklyn/ambari/rest/AmbariRestClient.java:56]: @ygy please remove static field.

File: ambari/src/main/java/io/brooklyn/ambari/rest/AmbariRestClient.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+    public AmbariRestClient() {
+        client = new UrlConnectionClient(); //Default Retrofit client
```
Comment (bostko) [inline ambari/src/main/java/io/brooklyn/ambari/rest/AmbariRestClient.java:59]: Why are you using `UrlConnectionClient`?

**Pull request merged**